homie-controller/examples/discover.rs

```diff
@@ -0,0 +1,37 @@
+use homie_controller::{Event, HomieController, PollError};
+use rumqttc::MqttOptions;
+
+#[tokio::main(core_threads = 2)]
+async fn main() -> Result<(), PollError> {
+    pretty_env_logger::init();
+
+    let mqttoptions = MqttOptions::new("homie_controller", "test.mosquitto.org", 1883);
+
+    let (controller, mut event_loop) = HomieController::new(mqttoptions, "homie");
+    controller.start().await?;
+    loop {
+        if let Some(event) = controller.poll(&mut event_loop).await? {
+            match event {
+                Event::PropertyValueChanged {
+                    device_id,
+                    node_id,
+                    property_id,
+                    value,
+                } => {
+                    println!("{}/{}/{} = {}", device_id, node_id, property_id, value);
+                }
+                _ => {
+                    println!("Event: {:?}", event);
+                    println!("Devices:");
+                    for device in controller.devices.lock().await.values() {
+                        if device.has_required_attributes() {
+                            println!(" * {:?}", device);
+                        } else {
+                            println!(" * {} not ready.", device.id);
+                        }
+                    }
```

Comment:
My preference would be for making the current `controller.devices` private and exposing a `controller.for_each_device(|device: & Device| {})` similar to https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.for_each (we can't implement Iterator directly, which is a bit annoying. I really wish someone would make a PushIterator/InternalIterator trait that we could use in times like these)

something like:
```suggestion
                    controller.for_each_device(|device| {
                        if device.has_required_attributes() {
                            println!(" * {:?}", device);
                        } else {
                            println!(" * {} not ready.", device.id);
                        }
                    })
```

When we come across use-cases that require other ways to access the device tree then I'm happy to iterate on it, but handing out mutable access to our internal state feels like the wrong answer.

In general, I would be in favour of iterating on our API design by building tools that solve real problems (I've been thinking about http://hintjens.com/blog:106 a bunch recently, and pondering how I can get closer to that ideal).

Reply:
I'd like to make it easy for users of the library not just to iterate over devices but also to look up specific ones. I agree that we shouldn't be handing out mutable access to them though, that is at least fixed now.